fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/RewriteSumByAssociativeRule.java

```diff
@@ -0,0 +1,340 @@
+                                ConstantOperator.createBigint(Long.parseLong(arg0.toString()));
+                        break;
+                    default:
+                        Preconditions.checkState(false, "unexpected sum function result type");
+                }
+                CallOperator newMultiply = new CallOperator("multiply", returnType,
+                        Lists.newArrayList(
+                                countOperator, constOperator));
+                return newMultiply;
+            } else {
+                // generate sum(arg0)
+
+                ColumnRefOperator newColumnRef;
+
+                if (arg0.isColumnRef() && oldPreAggProjections.containsKey(arg0)) {
+                    // if arg0 is a ColumnRef and exists in the oldPreAggProjections,
+                    // we don't need to create a new ColumnRef
+                    newColumnRef = (ColumnRefOperator) arg0;
+                } else if (commonArguments.containsKey(arg0)) {
+                    // if arg0 has been created by the previous rewriting,
+                    // we don't need to create a new ColumnRef
+                    newColumnRef = commonArguments.get(arg0);
+                } else {
+                    newColumnRef = columnRefFactory.create(arg0, arg0.getType(), arg0.isNullable());
+                    newPreAggProjections.put(newColumnRef, arg0);
+                    commonArguments.put(arg0, newColumnRef);
+                }
+
+                Type sumFunctionType = returnType;
+                if (returnType.isDecimalV3()) {
+                    // for decimal type, we should keep the result's scale same as the input's scale
+                    int argScale = ((ScalarType) arg0.getType()).getScalarScale();
+                    sumFunctionType = ScalarType.createDecimalV3Type(PrimitiveType.DECIMAL128, 38, argScale);
+                }
+
+                AggregateFunction sumFunction = AggregateFunction.createBuiltin(
+                        FunctionSet.SUM, Lists.newArrayList(arg0.getType()), sumFunctionType, sumFunctionType,
+                        false, true, false);
+
+                CallOperator newAggFunction = new CallOperator("sum",
+                        sumFunctionType, Lists.newArrayList(newColumnRef), sumFunction);
+
+                ColumnRefOperator newAggRef = columnRefFactory.create(
+                        newAggFunction, newAggFunction.getType(), true);
+                newAggregations.put(newAggRef, newAggFunction);
+                return newAggRef;
+            }
+        }
+
+        @Override
+        public ScalarOperator visitVariableReference(ColumnRefOperator op, Void context) {
+            if (oldAggregations.containsKey(op)) {
+                CallOperator aggFunction = oldAggregations.get(op);
+                if (aggFunction.isAggregate() && !aggFunction.isDistinct()
+                        && aggFunction.getFnName().equals("sum") &&
```

Comment:
can we support other agg functions, for examples:

avg(v+c) = avg(v)+c
avg (distinct v+c) = avg(distinct v) + c
max(v+c) = max(v) +c
min(v+c) = min(v) +c
count(v + c) = count(v)
count(distinct v+c) = count(distinct v)

Reply:
> can we support other agg functions, for examples:
> 
> avg(v+c) = avg(v)+c avg (distinct v+c) = avg(distinct v) + c max(v+c) = max(v) +c min(v+c) = min(v) +c count(v + c) = count(v) count(distinct v+c) = count(distinct v)

yeah…I think we can rewrite them by this idea, I will do it later.